packages/medusa/src/strategies/batch-jobs/product/import.ts

```diff
@@ -0,0 +1,732 @@
+    const transactionManager = this.transactionManager_ ?? this.manager_
+    const productOps = await this.downloadImportOpsFile(
+      batchJobId,
+      OperationType.ProductCreate
+    )
+
+    for (const productOp of productOps) {
+      try {
+        await this.productService_
+          .withTransaction(transactionManager)
+          .create(
+            transformProductData(productOp) as unknown as CreateProductInput
+          )
+      } catch (e) {
+        ProductImportStrategy.throwDescriptiveError(productOp)
+      }
+
+      this.updateProgress(batchJobId)
+    }
+  }
+
+  /**
+   * Method updates existing products in the DB using a CSV row data.
+   *
+   * @param batchJobId - An id of the current batch job being processed.
+   */
+  private async updateProducts(batchJobId: string): Promise<void> {
+    const transactionManager = this.transactionManager_ ?? this.manager_
+    const productOps = await this.downloadImportOpsFile(
+      batchJobId,
+      OperationType.ProductUpdate
+    )
+
+    for (const productOp of productOps) {
+      try {
+        await this.productService_
+          .withTransaction(transactionManager)
+          .update(
+            productOp["product.id"] as string,
+            transformProductData(productOp)
+          )
+      } catch (e) {
+        ProductImportStrategy.throwDescriptiveError(productOp)
+      }
+
+      this.updateProgress(batchJobId)
+    }
+  }
+
+  /**
+   * Method creates product variants from a CSV data.
+   * Method also handles processing of variant options.
+   *
+   * @param batchJobId - An id of the current batch job being processed.
+   */
+  private async createVariants(batchJobId: string): Promise<void> {
+    const transactionManager = this.transactionManager_ ?? this.manager_
+
+    const variantOps = await this.downloadImportOpsFile(
+      batchJobId,
+      OperationType.VariantCreate
+    )
+
+    for (const variantOp of variantOps) {
+      try {
+        const variant = transformVariantData(variantOp)
+
+        const product = await this.productService_
+          .withTransaction(transactionManager)
+          .retrieveByHandle(variantOp["product.handle"] as string, {
+            relations: ["variants", "variants.options", "options"],
+          })
+
+        const optionIds =
+          (variantOp["product.options"] as Record<string, string>[])?.map(
+            (variantOption) =>
+              product!.options.find(
+                (createdProductOption) =>
+                  createdProductOption.title === variantOption.title
+              )!.id
+          ) || []
+
+        variant.options =
+          (variant.options as object[])?.map((o, index) => ({
+            ...o,
+            option_id: optionIds[index],
+          })) || []
+
+        await this.productVariantService_
+          .withTransaction(transactionManager)
+          .create(product!, variant as unknown as CreateProductVariantInput)
+
+        this.updateProgress(batchJobId)
```

Comment:
todo: await, i will not comment the next ones 😅

Reply:
`updateProgress` & `finalize` were intentionally called without `await`. Now I think that `finalize` should  be awaited but `updateProgress` can be left as is, i.e. we can update progress info without blocking the working loop. wdyt?

Reply:
Progress was initially intended to inform the user about the general progression of a job. It is desynchronised right now in a way since we update progress after every 100 processed rows. Then it really doesn't matter if we `await` for the method but in return, we get non blocking of the main working loop